### PR TITLE
Fix haddocks on infix constructors

### LINKED ISSUE #758: AST differs on Haddock comment on infix constructor
**Describe the bug**
This snippet can't be formatted without `--unsafe` (minimized from [Agda](https://github.com/agda/agda/blob/3376fcd27edebfa90a3267cb26187f4bc6318c96/src/full/Agda/Syntax/Internal.hs#L641-L647)):

```haskell
data A = A :# A -- ^ foo
```

With `--unsafe`, it is formatted to
```haskell
data A
  = -- | foo
    A :# A
```

In the first case, the Haddock comment is attached to the constructor `:#`, but in the second case, it is attached to the first argument of `:#`.

**Expected behavior**
It could be formatted like this to preserve the AST:

```haskell
data A
  = A
      -- | foo
      :# A
```
But this is not really nice from a sylistic perspective (we should not have to break apart an existing single line constructor variant).

**Environment**
Ormolu 0.2.0.0

**Additional context**
This changed in GHC 9.0; previously, both snippets had the same AST.

### PR DESCRIPTION
**Pull request by brandonchinn178**
Resolves #758. Recommend reviewing commit-by-commit

### CODE REVIEW

**Review by amesgen — APPROVED** [review]
Thanks!

While the refactorings in the first three commits don't seem strictly necessary, they are nicely separated and make sense in the context of the additional `where` bindings introduced later.

Minor: I usually review changes that involve formatting with [difftastic](https://github.com/Wilfred/difftastic), so I would even be fine if you didn't separate out the formatting commits, but it certainly doesn't hurt; your efforts to increase reviewability are really appreciated!

**Comment by brandonchinn178** [comment]
I always organize my commits anyway, because it's easier for me to reorganize and edit my own work 😃

File: src/Ormolu/Printer/Meat/Declaration/Data.hs [hunk 1]
```diff
@@ -178,15 +177,14 @@ p_conDecl singleConstRec ConDeclH98 {..} = do
       breakpoint
     forM_ con_mb_cxt p_lhsContext
   switchLayout conDeclSpn $ case con_args of
-    PrefixCon [] xs -> do
```
[the remark this answers is not in the document]
Comment (brandonchinn178) [reply src/Ormolu/Printer/Meat/Declaration/Data.hs]: Makes sense. It is unrelated, but I was getting very confused about there being 4 branches for 3 constructors. Let me see if I can bring back the future-proofing though

[the remark this answers is not in the document]
Comment (brandonchinn178) [reply src/Ormolu/Printer/Meat/Declaration/Data.hs]: @amesgen what do you think about this? it keeps the 1-branch-per-constructor, but still checks for Void. https://github.com/tweag/ormolu/pull/1118/commits/622ee9835312f7c3f53d73825039362885b1e7da

[the remark this answers is not in the document]
Comment (amesgen) [reply src/Ormolu/Printer/Meat/Declaration/Data.hs]: Yes, that looks good; or now that I think of it, using a pattern like `PrefixCon (_ :: [Void]) xs` should also work.

**Pull request merged**